core/src/stored/backends/crud_storage.cc

```diff
@@ -0,0 +1,362 @@
+  std::string cmdline
+      = fmt::format("\"{}\" stat \"{}\" \"{}\"", m_program, obj_name, obj_part);
+  auto bph{BPipeHandle(cmdline.c_str(), m_program_timeout, "r", m_env_vars)};
+  auto rfh = bph.getReadFd();
+  Stat stat;
+  if (int n = fscanf(rfh, "%zu\n", &stat.size); n != 1) {
+    return tl::unexpected(
+        fmt::format("could not parse data returned by {}\n", cmdline));
+  }
+  if (auto ret = bph.close(); ret != 0) {
+    Dmsg1(debug_info, "stat returned %d\n", ret);
+    return tl::unexpected(
+        fmt::format("Running \"{}\" returned {}\n", cmdline, ret));
+  }
+  Dmsg1(debug_trace, "stat returns %zu\n", stat.size);
+  return stat;
+}
+
+auto CrudStorage::list(std::string_view obj_name)
+    -> tl::expected<std::map<std::string, Stat>, std::string>
+{
+  Dmsg1(debug_trace, "list %s called\n", obj_name.data());
+  std::string cmdline = fmt::format("\"{}\" list \"{}\"", m_program, obj_name);
+  auto bph{BPipeHandle(cmdline.c_str(), m_program_timeout, "r", m_env_vars)};
+  auto rfh = bph.getReadFd();
+
+  std::map<std::string, Stat> result;
+  while (!feof(rfh)) {
+    Stat stat;
+    auto obj_part = std::string(128, '\0');
+    if (int n = fscanf(rfh, "%128s %zu\n", obj_part.data(), &stat.size);
+        n != 2) {
+      Dmsg1(debug_info, "fscanf() returned %d\n", n);
+      return tl::unexpected(
+          fmt::format("could not parse data returned by {}", cmdline));
+    }
+    obj_part.resize(std::strlen(obj_part.c_str()));
+    result[obj_part] = stat;
+
+    Dmsg1(debug_trace, "volume=%s part=%s size=%zu\n", obj_name.data(),
+          obj_part.c_str(), stat.size);
+  }
+
+  if (auto ret = bph.close(); ret != 0) {
+    Dmsg1(debug_info, "list returned %d\n", ret);
+    return tl::unexpected(
+        fmt::format("Running \"{}\" returned {}\n", cmdline, ret));
+  }
+  return result;
+}
+
+tl::expected<void, std::string> CrudStorage::upload(std::string_view obj_name,
+                                                    std::string_view obj_part,
+                                                    gsl::span<char> obj_data)
+{
+  Dmsg1(debug_trace, "upload %s/%s called\n", obj_name.data(), obj_part.data());
+  std::string cmdline = fmt::format("\"{}\" upload \"{}\" \"{}\"", m_program,
+                                    obj_name, obj_part);
+
+  auto bph{BPipeHandle(cmdline.c_str(), m_program_timeout, "rw", m_env_vars)};
+  auto wfh = bph.getWriteFd();
+
+  constexpr size_t max_write_size{256 * 1024};
+  size_t remaining_bytes{obj_data.size()};
+
+  while (remaining_bytes > 0) {
+    const size_t write_size = std::min(max_write_size, remaining_bytes);
+    const size_t offset = obj_data.size() - remaining_bytes;
+    if (auto has_written = fwrite(obj_data.data() + offset, 1, write_size, wfh);
+        has_written != write_size) {
+      return tl::unexpected(fmt::format(
+          "Broken pipe after writing {} of {} bytes at offset {} into {}/{}\n",
+          has_written, write_size, offset, obj_name, obj_part));
+    }
```

Comment:
`has_written != write_size` is always an error condition:
> The fwrite() function shall return the number of elements successfully written, which shall be less than nitems only if a write error is encountered.

I agree that it could be `EINTR` or even `EAGAIN`, so restarting the call probably makes sense here.